diff --git a/Dockerfile b/Dockerfile
index c626a5f..d730600 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -1,8 +1,9 @@
-FROM ruby:2.5
+FROM ruby:2.6.6
 RUN apt-get update -qq && apt-get install -y build-essential libpq-dev nodejs
 RUN mkdir /code
 WORKDIR /code
 COPY Gemfile /code/Gemfile
 COPY Gemfile.lock /code/Gemfile.lock
+RUN bundle update rails
 RUN bundle install
 COPY . /code
diff --git a/Gemfile b/Gemfile
index 5bc09a6..049874c 100644
--- a/Gemfile
+++ b/Gemfile
@@ -1,7 +1,7 @@
 source 'https://rubygems.org'
 git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 
-ruby '2.5.1'
+ruby '2.6.6'
 
 # Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
 gem 'rails', '~> 5.2.3'
diff --git a/Gemfile.lock b/Gemfile.lock
index 81de9c3..e51ee98 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -1,67 +1,66 @@
 GEM
   remote: https://rubygems.org/
   specs:
-    actioncable (5.2.3)
-      actionpack (= 5.2.3)
+    actioncable (5.2.4.5)
+      actionpack (= 5.2.4.5)
       nio4r (~> 2.0)
       websocket-driver (>= 0.6.1)
-    actionmailer (5.2.3)
-      actionpack (= 5.2.3)
-      actionview (= 5.2.3)
-      activejob (= 5.2.3)
+    actionmailer (5.2.4.5)
+      actionpack (= 5.2.4.5)
+      actionview (= 5.2.4.5)
+      activejob (= 5.2.4.5)
       mail (~> 2.5, >= 2.5.4)
       rails-dom-testing (~> 2.0)
-    actionpack (5.2.3)
-      actionview (= 5.2.3)
-      activesupport (= 5.2.3)
-      rack (~> 2.0)
+    actionpack (5.2.4.5)
+      actionview (= 5.2.4.5)
+      activesupport (= 5.2.4.5)
+      rack (~> 2.0, >= 2.0.8)
       rack-test (>= 0.6.3)
       rails-dom-testing (~> 2.0)
       rails-html-sanitizer (~> 1.0, >= 1.0.2)
-    actionview (5.2.3)
-      activesupport (= 5.2.3)
+    actionview (5.2.4.5)
+      activesupport (= 5.2.4.5)
       builder (~> 3.1)
       erubi (~> 1.4)
       rails-dom-testing (~> 2.0)
       rails-html-sanitizer (~> 1.0, >= 1.0.3)
-    activejob (5.2.3)
-      activesupport (= 5.2.3)
+    activejob (5.2.4.5)
+      activesupport (= 5.2.4.5)
       globalid (>= 0.3.6)
-    activemodel (5.2.3)
-      activesupport (= 5.2.3)
-    activerecord (5.2.3)
-      activemodel (= 5.2.3)
-      activesupport (= 5.2.3)
+    activemodel (5.2.4.5)
+      activesupport (= 5.2.4.5)
+    activerecord (5.2.4.5)
+      activemodel (= 5.2.4.5)
+      activesupport (= 5.2.4.5)
       arel (>= 9.0)
-    activestorage (5.2.3)
-      actionpack (= 5.2.3)
-      activerecord (= 5.2.3)
+    activestorage (5.2.4.5)
+      actionpack (= 5.2.4.5)
+      activerecord (= 5.2.4.5)
       marcel (~> 0.3.1)
-    activesupport (5.2.3)
+    activesupport (5.2.4.5)
       concurrent-ruby (~> 1.0, >= 1.0.2)
       i18n (>= 0.7, < 2)
       minitest (~> 5.1)
       tzinfo (~> 1.1)
-    addressable (2.6.0)
-      public_suffix (>= 2.0.2, < 4.0)
+    addressable (2.7.0)
+      public_suffix (>= 2.0.2, < 5.0)
     archive-zip (0.12.0)
       io-like (~> 0.3.0)
     arel (9.0.0)
-    bindex (0.6.0)
-    bootsnap (1.4.2)
+    bindex (0.8.1)
+    bootsnap (1.7.2)
       msgpack (~> 1.0)
-    builder (3.2.3)
-    byebug (11.0.1)
-    capybara (3.16.1)
+    builder (3.2.4)
+    byebug (11.1.3)
+    capybara (3.35.3)
       addressable
       mini_mime (>= 0.1.3)
       nokogiri (~> 1.8)
       rack (>= 1.6.0)
       rack-test (>= 0.6.3)
-      regexp_parser (~> 1.2)
+      regexp_parser (>= 1.5, < 3.0)
       xpath (~> 3.2)
-    childprocess (0.9.0)
-      ffi (~> 1.0, >= 1.0.11)
+    childprocess (3.0.0)
     chromedriver-helper (2.1.1)
       archive-zip (~> 0.10)
       nokogiri (~> 1.8)
@@ -72,132 +71,136 @@ GEM
       coffee-script-source
       execjs
     coffee-script-source (1.12.2)
-    concurrent-ruby (1.1.5)
-    crass (1.0.4)
-    erubi (1.8.0)
+    concurrent-ruby (1.1.8)
+    crass (1.0.6)
+    erubi (1.10.0)
     execjs (2.7.0)
-    faraday (0.15.4)
+    faraday (1.3.0)
+      faraday-net_http (~> 1.0)
       multipart-post (>= 1.2, < 3)
-    ffi (1.10.0)
+      ruby2_keywords
+    faraday-net_http (1.0.1)
+    ffi (1.14.2)
     globalid (0.4.2)
       activesupport (>= 4.2.0)
-    i18n (1.6.0)
+    i18n (1.8.9)
       concurrent-ruby (~> 1.0)
-    io-like (0.3.0)
-    jbuilder (2.8.0)
-      activesupport (>= 4.2.0)
-      multi_json (>= 1.2)
-    jwt (2.1.0)
+    io-like (0.3.1)
+    jbuilder (2.11.2)
+      activesupport (>= 5.0.0)
+    jwt (2.2.2)
     listen (3.1.5)
       rb-fsevent (~> 0.9, >= 0.9.4)
       rb-inotify (~> 0.9, >= 0.9.7)
       ruby_dep (~> 1.2)
-    loofah (2.2.3)
+    loofah (2.9.0)
       crass (~> 1.0.2)
       nokogiri (>= 1.5.9)
     mail (2.7.1)
       mini_mime (>= 0.1.1)
     marcel (0.3.3)
       mimemagic (~> 0.3.2)
-    method_source (0.9.2)
-    mimemagic (0.3.3)
-    mini_mime (1.0.1)
-    mini_portile2 (2.4.0)
-    minitest (5.11.3)
-    msgpack (1.2.9)
-    multi_json (1.13.1)
+    method_source (1.0.0)
+    mimemagic (0.3.5)
+    mini_mime (1.0.2)
+    mini_portile2 (2.5.0)
+    minitest (5.14.3)
+    msgpack (1.4.2)
+    multi_json (1.15.0)
     multi_xml (0.6.0)
-    multipart-post (2.0.0)
-    nio4r (2.3.1)
-    nokogiri (1.10.2)
-      mini_portile2 (~> 2.4.0)
-    oauth2 (1.4.1)
-      faraday (>= 0.8, < 0.16.0)
+    multipart-post (2.1.1)
+    nio4r (2.5.5)
+    nokogiri (1.11.1)
+      mini_portile2 (~> 2.5.0)
+      racc (~> 1.4)
+    oauth2 (1.4.4)
+      faraday (>= 0.8, < 2.0)
       jwt (>= 1.0, < 3.0)
       multi_json (~> 1.3)
       multi_xml (~> 0.5)
       rack (>= 1.2, < 3)
-    public_suffix (3.0.3)
-    puma (3.12.1)
-    rack (2.0.7)
+    public_suffix (4.0.6)
+    puma (3.12.6)
+    racc (1.5.2)
+    rack (2.2.3)
     rack-test (1.1.0)
       rack (>= 1.0, < 3)
-    rails (5.2.3)
-      actioncable (= 5.2.3)
-      actionmailer (= 5.2.3)
-      actionpack (= 5.2.3)
-      actionview (= 5.2.3)
-      activejob (= 5.2.3)
-      activemodel (= 5.2.3)
-      activerecord (= 5.2.3)
-      activestorage (= 5.2.3)
-      activesupport (= 5.2.3)
+    rails (5.2.4.5)
+      actioncable (= 5.2.4.5)
+      actionmailer (= 5.2.4.5)
+      actionpack (= 5.2.4.5)
+      actionview (= 5.2.4.5)
+      activejob (= 5.2.4.5)
+      activemodel (= 5.2.4.5)
+      activerecord (= 5.2.4.5)
+      activestorage (= 5.2.4.5)
+      activesupport (= 5.2.4.5)
       bundler (>= 1.3.0)
-      railties (= 5.2.3)
-      sprockets-rails (>= 2.0.0)
+      railties (= 5.2.4.5)
+      sprockets-rails (>= 3.0.0)
     rails-dom-testing (2.0.3)
       activesupport (>= 4.2.0)
       nokogiri (>= 1.6)
-    rails-html-sanitizer (1.0.4)
-      loofah (~> 2.2, >= 2.2.2)
-    railties (5.2.3)
-      actionpack (= 5.2.3)
-      activesupport (= 5.2.3)
+    rails-html-sanitizer (1.3.0)
+      loofah (~> 2.3)
+    railties (5.2.4.5)
+      actionpack (= 5.2.4.5)
+      activesupport (= 5.2.4.5)
       method_source
       rake (>= 0.8.7)
       thor (>= 0.19.0, < 2.0)
-    rake (12.3.2)
-    rb-fsevent (0.10.3)
-    rb-inotify (0.10.0)
+    rake (13.0.3)
+    rb-fsevent (0.10.4)
+    rb-inotify (0.10.1)
       ffi (~> 1.0)
-    regexp_parser (1.4.0)
+    regexp_parser (2.0.3)
+    ruby2_keywords (0.0.4)
     ruby_dep (1.5.0)
-    rubyzip (1.2.2)
+    rubyzip (2.3.0)
     sass (3.7.4)
       sass-listen (~> 4.0.0)
     sass-listen (4.0.0)
       rb-fsevent (~> 0.9, >= 0.9.4)
       rb-inotify (~> 0.9, >= 0.9.7)
-    sass-rails (5.0.7)
-      railties (>= 4.0.0, < 6)
+    sass-rails (5.1.0)
+      railties (>= 5.2.0)
       sass (~> 3.1)
       sprockets (>= 2.8, < 4.0)
       sprockets-rails (>= 2.0, < 4.0)
       tilt (>= 1.1, < 3)
-    selenium-webdriver (3.141.0)
-      childprocess (~> 0.5)
-      rubyzip (~> 1.2, >= 1.2.2)
-    spring (2.0.2)
-      activesupport (>= 4.2)
+    selenium-webdriver (3.142.7)
+      childprocess (>= 0.5, < 4.0)
+      rubyzip (>= 1.2.2)
+    spring (2.1.1)
     spring-watcher-listen (2.0.1)
       listen (>= 2.7, < 4.0)
       spring (>= 1.2, < 3.0)
     sprockets (3.7.2)
       concurrent-ruby (~> 1.0)
       rack (> 1, < 3)
-    sprockets-rails (3.2.1)
+    sprockets-rails (3.2.2)
       actionpack (>= 4.0)
       activesupport (>= 4.0)
       sprockets (>= 3.0.0)
-    sqlite3 (1.4.0)
-    thor (0.20.3)
+    sqlite3 (1.4.2)
+    thor (1.1.0)
     thread_safe (0.3.6)
-    tilt (2.0.9)
-    turbolinks (5.2.0)
+    tilt (2.0.10)
+    turbolinks (5.2.1)
       turbolinks-source (~> 5.2)
     turbolinks-source (5.2.0)
-    tzinfo (1.2.5)
+    tzinfo (1.2.9)
       thread_safe (~> 0.1)
-    uglifier (4.1.20)
+    uglifier (4.2.0)
       execjs (>= 0.3.0, < 3)
     web-console (3.7.0)
       actionview (>= 5.0)
       activemodel (>= 5.0)
       bindex (>= 0.4.0)
       railties (>= 5.0)
-    websocket-driver (0.7.0)
+    websocket-driver (0.7.3)
       websocket-extensions (>= 0.1.0)
-    websocket-extensions (0.1.3)
+    websocket-extensions (0.1.5)
     xpath (3.2.0)
       nokogiri (~> 1.8)
 
@@ -205,6 +208,9 @@ PLATFORMS
   ruby
 
 DEPENDENCIES
+  actionview (>= 5.2.3)
+  activejob (>= 5.2.3)
+  activestorage (>= 5.2.3)
   bootsnap (>= 1.1.0)
   byebug
   capybara (>= 2.15, < 4.0)
@@ -212,9 +218,13 @@ DEPENDENCIES
   coffee-rails (~> 4.2.2)
   jbuilder (~> 2.5)
   listen (>= 3.0.5, < 3.2)
+  loofah (>= 2.2.3)
+  nokogiri (>= 1.10.2)
   oauth2 (~> 1.2)
   puma (~> 3.11)
-  rails (~> 5.2.0)
+  rack (>= 2.0.7)
+  rails (~> 5.2.3)
+  railties (>= 5.2.3)
   sass-rails (~> 5.0)
   selenium-webdriver
   spring
@@ -226,7 +236,7 @@ DEPENDENCIES
   web-console (>= 3.3.0)
 
 RUBY VERSION
-   ruby 2.5.1p57
+   ruby 2.6.4p104
 
 BUNDLED WITH
-   1.16.3
+   1.17.3
diff --git a/app/controllers/bluebutton/resources/resources_controller.rb b/app/controllers/bluebutton/resources/resources_controller.rb
index 47af0c0..5e63a70 100644
--- a/app/controllers/bluebutton/resources/resources_controller.rb
+++ b/app/controllers/bluebutton/resources/resources_controller.rb
@@ -9,42 +9,47 @@ class Bluebutton::Resources::ResourcesController < ApplicationController
 
   def userinfo
     resource_title('User Info')
-    @bb_response = @bluebutton_oauth_service.get_resource('/v1/connect/userinfo')
+    @bb_response = @bluebutton_oauth_service.get_resource('/' + @bluebutton_oauth_service.api_ver + '/connect/userinfo')
     render('result')
   end
 
   def patient
     resource_title('Patient Information')
-    @bb_response = @bluebutton_oauth_service.get_resource('/v1/fhir/Patient/')
+    @bb_response = @bluebutton_oauth_service.get_resource('/' + @bluebutton_oauth_service.api_ver + '/fhir/Patient/')
     render('result')
   end
 
   def eob
     resource_title('Explanation Of Benefit')
-    @bb_response = @bluebutton_oauth_service.get_resource('/v1/fhir/ExplanationOfBenefit/')
+    @bb_response = @bluebutton_oauth_service.get_resource('/' + @bluebutton_oauth_service.api_ver + '/fhir/ExplanationOfBenefit/')
     render('result')
   end
 
   def coverage
     resource_title('Coverage')
-    @bb_response = @bluebutton_oauth_service.get_resource('/v1/fhir/Coverage/')
+    @bb_response = @bluebutton_oauth_service.get_resource('/' + @bluebutton_oauth_service.api_ver + '/fhir/Coverage/')
     render('result')
   end
 
   def oidc
     resource_title('OIDC Discovery')
-    @bb_response = @bluebutton_oauth_service.get_resource('/.well-known/openid-configuration')
+    if @bluebutton_oauth_service.api_ver == 'v1'
+      @bb_response = @bluebutton_oauth_service.get_resource('/.well-known/openid-configuration')
+    else
+      @bb_response = @bluebutton_oauth_service.get_resource('/.well-known/openid-configuration-v2')
+    end
     render('result')
   end
 
   def fhirmeta
     resource_title('FHIR Metadata')
-    @bb_response = @bluebutton_oauth_service.get_resource('/v1/fhir/metadata')
+    @bb_response = @bluebutton_oauth_service.get_resource('/' + @bluebutton_oauth_service.api_ver + '/fhir/metadata')
     render('result')
   end
 
   def get_resource
     resource_title(params[:resource_title])
+    p(params[:page_url])
     @bb_response = @bluebutton_oauth_service.get_resource(params[:page_url])
     render('result')
   end
diff --git a/app/services/bluebutton_oauth_service.rb b/app/services/bluebutton_oauth_service.rb
index 3c01e69..f680e90 100644
--- a/app/services/bluebutton_oauth_service.rb
+++ b/app/services/bluebutton_oauth_service.rb
@@ -5,16 +5,17 @@ class BluebuttonOauthService
   BB_CLIENT_ID = LOCAL_ENV['BB_CLIENT_ID']
   BB_CLIENT_SECRET = LOCAL_ENV['BB_CLIENT_SECRET']
   BB_REDIRECT_URI = LOCAL_ENV['BB_REDIRECT_URI']
-  BB_AUTHORIZE_URL = LOCAL_ENV['BB_AUTHORIZE_URL']
-  BB_TOKEN_URL = LOCAL_ENV['BB_TOKEN_URL']
   BB_BASE_URL = LOCAL_ENV['BB_BASE_URL']
-  BB_REFER_URL = 'http://localhost:3000/bluebutton_authorize'
+  BB_AUTHORIZE_URL = BB_BASE_URL + '/' + LOCAL_ENV['BB_API_VER'] + LOCAL_ENV['BB_AUTHORIZE_URI']
+  BB_TOKEN_URL = BB_BASE_URL + '/' + LOCAL_ENV['BB_API_VER'] + LOCAL_ENV['BB_TOKEN_URI']
+  BB_REFER_URL = 'http://192.168.0.129:3000/bluebutton_authorize'
 
-  attr_accessor :bb_acc_token, :state 
+  attr_accessor :bb_acc_token, :state, :api_ver
 
   def initialize(bb_acc_token, state)
     @bb_acc_token = bb_acc_token
     @state = state
+    @api_ver = LOCAL_ENV['BB_API_VER']
     @client = OAuth2::Client.new(BB_CLIENT_ID, BB_CLIENT_SECRET, {
       :authorize_url => BB_AUTHORIZE_URL,
       :token_url => BB_TOKEN_URL,
@@ -80,6 +81,10 @@ class BluebuttonOauthService
     nil
   end
 
+  def api_ver
+    @api_ver
+  end
+
   # Resource method:
   def get_resource(url)
     response_json = @access_token.get(url).body
diff --git a/app/views/bluebutton/demo/authorize.html.erb b/app/views/bluebutton/demo/authorize.html.erb
index c6f0d11..2104c6f 100644
--- a/app/views/bluebutton/demo/authorize.html.erb
+++ b/app/views/bluebutton/demo/authorize.html.erb
@@ -2,6 +2,11 @@
 
 <h2>Authorize This Application </h2>
 
+<h2 style="color:red"><strong> API version: <%= @bluebutton_oauth_service.api_ver %></strong></h2>
+<p>BB_URL_PREFIX in settings (e.g. local_env.yml) indicates which API versions to use,<br>
+BB_URL_PREFIX: "v1" - use v1 API, "v2" - use v2 API, to switch, shutdown sample client,<br>
+change the setting, then start again.
+</p>
 <h3>The following link will take you to the MyMedicare site to enter your login information and authorize this application to access your data.</h3>
 
 <h2><%= link_to 'Click here to authorize!',  @bluebutton_oauth_service.get_authorize_url %></h2>
diff --git a/app/views/bluebutton/demo/main.html.erb b/app/views/bluebutton/demo/main.html.erb
index d07d0f7..bb38942 100644
--- a/app/views/bluebutton/demo/main.html.erb
+++ b/app/views/bluebutton/demo/main.html.erb
@@ -1,5 +1,7 @@
 <h1> <b>Blue Button API Resources</b></h1>
 
+<h2 style="color:red"><strong> API version: <%= @bluebutton_oauth_service.api_ver %></strong></h2>
+
 <h2><u>Patient ID:  <%= session[:bb_patient_id] %> </u></h2>
 
 <h3><%= link_to 'Click to return to the authorization page.',  root_path %></h3>
diff --git a/app/views/bluebutton/resources/resources/result.html.erb b/app/views/bluebutton/resources/resources/result.html.erb
index 7e9e0a7..03a338a 100644
--- a/app/views/bluebutton/resources/resources/result.html.erb
+++ b/app/views/bluebutton/resources/resources/result.html.erb
@@ -1,5 +1,7 @@
 <h1> <%= @resource_title %></h1>
 
+<h2 style="color:red"><strong> API version: <%= @bluebutton_oauth_service.api_ver %></strong></h2>
+
 <h2><u> Patient ID: <%= session[:bb_patient_id] %></u></h2>
 
 <h3><%= link_to 'Click to return to the resources demo page.',  bluebutton_main_path %></h3>
diff --git a/config/initializers/assets.rb b/config/initializers/assets.rb
index 4b828e8..a22ec6d 100644
--- a/config/initializers/assets.rb
+++ b/config/initializers/assets.rb
@@ -12,3 +12,11 @@ Rails.application.config.assets.paths << Rails.root.join('node_modules')
 # application.js, application.css, and all non-JS/CSS in the app/assets
 # folder are already added.
 # Rails.application.config.assets.precompile += %w( admin.js admin.css )
+
+Rails.application.config.assets.configure do |env|
+    env.cache = Sprockets::Cache::FileStore.new(
+      ENV.fetch("SPROCKETS_CACHE", "#{env.root}/tmp/cache/assets"),
+      Rails.application.config.assets.cache_limit,
+      env.logger
+    )
+end
\ No newline at end of file
diff --git a/config/local_env_sample.yml b/config/local_env_sample.yml
index 9f8886c..0826ed3 100644
--- a/config/local_env_sample.yml
+++ b/config/local_env_sample.yml
@@ -8,12 +8,17 @@ development:
   BB_CLIENT_SECRET: "<enter client secret here>"
   # Callback handler URL:
   BB_REDIRECT_URI: "http://localhost:3000/bluebutton/callback"
+  #BB_API_VER: "v2"
+  BB_API_VER: "v1"
   # Authorization endpoint:
-  BB_AUTHORIZE_URL: "https://sandbox.bluebutton.cms.gov/v1/o/authorize/"
+  BB_AUTHORIZE_URI: "/o/authorize/"
+  # BB_AUTHORIZE_URL: "https://sandbox.bluebutton.cms.gov/v1/o/authorize/"
   # Exchange code for Token endpoint:
-  BB_TOKEN_URL: "https://sandbox.bluebutton.cms.gov/v1/o/token/"
+  BB_TOKEN_URI: "/o/token/"
+  # BB_TOKEN_URL: "https://sandbox.bluebutton.cms.gov/v1/o/token/"
   # Base Bluebutton site path:
-  BB_BASE_URL: "https://sandbox.bluebutton.cms.gov"
+  # BB_BASE_URL: "https://sandbox.bluebutton.cms.gov"
+  BB_BASE_URL: "http://192.168.0.109:8000"
 
 production:
   BB_CLIENT_ID: "<enter client id here>"
diff --git a/docker-compose.env b/docker-compose.env
index b7a89fd..d81454c 100644
--- a/docker-compose.env
+++ b/docker-compose.env
@@ -1,2 +1,3 @@
 # will log both request and response, including bodies
 OAUTH_DEBUG="true"
+SPROCKETS_CACHE=/cache
\ No newline at end of file
diff --git a/docker-compose.yml b/docker-compose.yml
index 8ef8a9f..8aa0c0e 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -1,12 +1,16 @@
 version: '3'
 
-services:
 
+volumes:
+  cache: # Added to support Windows 10 Docker
+
+services:
   web:
     build: .
     command: bundle exec rails s -p 3000 -b '0.0.0.0'
     volumes:
       - .:/code
+      - cache:/cache # Added to support Windows 10 Docker
     ports:
       - "3000:3000"
     env_file:
