diff --git a/Gemfile b/Gemfile
index 5bc09a6..cfc6639 100644
--- a/Gemfile
+++ b/Gemfile
@@ -1,7 +1,7 @@
 source 'https://rubygems.org'
 git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 
-ruby '2.5.1'
+ruby '2.5.8'
 
 # Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
 gem 'rails', '~> 5.2.3'
diff --git a/Gemfile.lock b/Gemfile.lock
index 81de9c3..7f6a238 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -205,6 +205,9 @@ PLATFORMS
   ruby
 
 DEPENDENCIES
+  actionview (>= 5.2.3)
+  activejob (>= 5.2.3)
+  activestorage (>= 5.2.3)
   bootsnap (>= 1.1.0)
   byebug
   capybara (>= 2.15, < 4.0)
@@ -212,9 +215,13 @@ DEPENDENCIES
   coffee-rails (~> 4.2.2)
   jbuilder (~> 2.5)
   listen (>= 3.0.5, < 3.2)
+  loofah (>= 2.2.3)
+  nokogiri (>= 1.10.2)
   oauth2 (~> 1.2)
   puma (~> 3.11)
-  rails (~> 5.2.0)
+  rack (>= 2.0.7)
+  rails (~> 5.2.3)
+  railties (>= 5.2.3)
   sass-rails (~> 5.0)
   selenium-webdriver
   spring
@@ -226,7 +233,7 @@ DEPENDENCIES
   web-console (>= 3.3.0)
 
 RUBY VERSION
-   ruby 2.5.1p57
+   ruby 2.5.8p224
 
 BUNDLED WITH
-   1.16.3
+   1.17.3
diff --git a/app/controllers/bluebutton/resources/resources_controller.rb b/app/controllers/bluebutton/resources/resources_controller.rb
index 47af0c0..f20fe03 100644
--- a/app/controllers/bluebutton/resources/resources_controller.rb
+++ b/app/controllers/bluebutton/resources/resources_controller.rb
@@ -19,18 +19,36 @@ class Bluebutton::Resources::ResourcesController < ApplicationController
     render('result')
   end
 
+  def patientV2
+    resource_title('PatientV2 Information')
+    @bb_response = @bluebutton_oauth_service.get_resource('/v1/fhir/Patient/?fhir_ver=r4')
+    render('result')
+  end
+
   def eob
     resource_title('Explanation Of Benefit')
     @bb_response = @bluebutton_oauth_service.get_resource('/v1/fhir/ExplanationOfBenefit/')
     render('result')
   end
 
+  def eobV2
+    resource_title('Explanation Of Benefit V2')
+    @bb_response = @bluebutton_oauth_service.get_resource('/v1/fhir/ExplanationOfBenefit/?fhir_ver=r4')
+    render('result')
+  end
+
   def coverage
     resource_title('Coverage')
     @bb_response = @bluebutton_oauth_service.get_resource('/v1/fhir/Coverage/')
     render('result')
   end
 
+  def coverageV2
+    resource_title('Coverage V2')
+    @bb_response = @bluebutton_oauth_service.get_resource('/v1/fhir/Coverage/?fhir_ver=r4')
+    render('result')
+  end
+
   def oidc
     resource_title('OIDC Discovery')
     @bb_response = @bluebutton_oauth_service.get_resource('/.well-known/openid-configuration')
@@ -43,6 +61,12 @@ class Bluebutton::Resources::ResourcesController < ApplicationController
     render('result')
   end
 
+  def fhirmetaV2
+    resource_title('FHIR Metadata V2')
+    @bb_response = @bluebutton_oauth_service.get_resource('/v1/fhir/metadata/?fhir_ver=r4')
+    render('result')
+  end
+
   def get_resource
     resource_title(params[:resource_title])
     @bb_response = @bluebutton_oauth_service.get_resource(params[:page_url])
diff --git a/app/views/bluebutton/demo/main.html.erb b/app/views/bluebutton/demo/main.html.erb
index d07d0f7..c5d643f 100644
--- a/app/views/bluebutton/demo/main.html.erb
+++ b/app/views/bluebutton/demo/main.html.erb
@@ -23,16 +23,31 @@
     <td><%= link_to 'Patient',  bluebutton_resources_patient_path %></td>
   </tr>
 
+  <tr>
+    <td>Detailed information for patient V2</td>
+    <td><%= link_to 'PatientV2',  bluebutton_resources_patientV2_path %></td>
+  </tr>
+
   <tr>
     <td>Explanation of benefit information for patient</td>
     <td><%= link_to 'ExplanationOfBenefit',  bluebutton_resources_eob_path %></td>
   </tr>
 
+  <tr>
+    <td>Explanation of benefit information for patient V2</td>
+    <td><%= link_to 'ExplanationOfBenefitV2',  bluebutton_resources_eobV2_path %></td>
+  </tr>
+
   <tr>
     <td>Coverage information for patient</td>
     <td><%= link_to 'Coverage',  bluebutton_resources_coverage_path %></td>
   </tr>
 
+  <tr>
+    <td>Coverage information for patient V2</td>
+    <td><%= link_to 'CoverageV2',  bluebutton_resources_coverageV2_path %></td>
+  </tr>
+
   <tr>
     <td>OIDC Discovery</td>
     <td><%= link_to 'OIDC_Disc',  bluebutton_resources_oidc_path %></td>
@@ -43,5 +58,9 @@
     <td><%= link_to 'FHIR_Meta',  bluebutton_resources_fhirmeta_path %></td>
   </tr>
 
+  <tr>
+    <td>FHIR Metadata</td>
+    <td><%= link_to 'FHIR_Meta V2',  bluebutton_resources_fhirmetaV2_path %></td>
+  </tr>
 </table>
 </h3>
diff --git a/config/routes.rb b/config/routes.rb
index 23dfc63..30a4ba0 100644
--- a/config/routes.rb
+++ b/config/routes.rb
@@ -20,18 +20,30 @@ Rails.application.routes.draw do
       # Bluebutton resource: patient
       get 'patient', to: 'resources#patient', as: 'patient'
 
-      # Bluebutton resource: eob = Explanation of Benefit 
+      # Bluebutton resource: patientV2
+      get 'patientV2', to: 'resources#patientV2', as: 'patientV2'
+
+      # Bluebutton resource: eob = Explanation of Benefit
       get 'eob', to: 'resources#eob', as: 'eob'
 
-      # Bluebutton resource: Coverage  
+      # Bluebutton resource: eobV2 = Explanation of Benefit V2
+      get 'eobV2', to: 'resources#eobV2', as: 'eobV2'
+
+      # Bluebutton resource: Coverage
       get 'coverage', to: 'resources#coverage', as: 'coverage'
 
+      # Bluebutton resource: CoverageV2
+      get 'coverageV2', to: 'resources#coverageV2', as: 'coverageV2'
+
       # Bluebutton resource: OIDC Discovery 
       get 'oidc', to: 'resources#oidc', as: 'oidc'
 
       # Bluebutton resource: FHIR Metadata 
       get 'fhirmeta', to: 'resources#fhirmeta', as: 'fhirmeta'
 
+      # Bluebutton resource: FHIR MetadataV2
+      get 'fhirmetaV2', to: 'resources#fhirmetaV2', as: 'fhirmetaV2'
+
       # Bluebutton resource: get_resource (for JSON responses with paging links)
       get 'get_resource', to: 'resources#get_resource', as: 'get_resource'
 
diff --git a/log/.keep b/log/.keep
deleted file mode 100644
index e69de29..0000000
diff --git a/tmp/.keep b/tmp/.keep
deleted file mode 100644
index e69de29..0000000
